Add API Server to lbry.go

tiger5226 opened the pull request:
moved api server from internal-apis to lbry.go so it can be reused.

lyoshenka commented:
In fact, a few things need to be fixed. We can’t use util.Debugging, we have to configure how errors get back to us (util.Send.ToSlack is too api-specific), we can’t hardcode lbry.io in the headers, etc

tiger5226 commented:
your right, best to just delete my local copy of internal-apis, to see what breaks. Then repackage that.

tiger5226 commented:
Made the following adjustments. 
-Removed dependency on internal apis
-Moved over only required packages.
-Adjusted slack.go to be generic instead of hardcoding channel name.

lyoshenka commented:
There's also slack code in lbryio/boardbot repo. It should be merged with this as well.

Also, I don't like that slack errors are ignored. They should be returned, and then we can ignore them when we call the SendToSlack function if we want. SendToSlack def shouldn't be logging the error itself.

tiger5226 commented:
- [x] move slack code from lbryio/boardbot

- [x] change package name to api

- [x] parameterize the header settings

- [x] remove git ignore

- [x] don't ignore slack errors, return them

- [x] strict transport security... parameterize

- [ ] util.Debugging solution

tiger5226 commented:
-Renamed package to api
-removed util.Debugging from server.go
-Added an ErrorHandling function that be used as interface for slack for internal-apis
-Added Map for Header settings that can be set before the serving
-Merged slack code from lbryio/boardbot
-Cleaned up the slack.go code so it made more sense and flowed better
-Removed gitignore entry for `.idea`, should be global
-Removed debugging.go

tiger5226 commented:
FYI - I am using the apiserver branch for chainquery until this is resolved. Please do not delete the branch. I will delete it once I correct the vendoring to use master.

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,299 @@
+// ie. "Strict-Transport-Security" - "max-age=31536000; preload"
+// ie. "Access-Control-Allow-Origin" -"<header.Origin>"
+// ie. "Access-Control-Allow-Methods" - "GET, POST, OPTIONS"
+var HeaderSettings *map[string]string
```
why is this a pointer? an uninitialized map already defaults to `nil`

tiger5226 replied:
good catch!

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,299 @@
+	}, "", "  ")
+	if err != nil {
+		errMsg := "Error encoding JSON response: " + err.Error()
+		log.Errorln(errMsg)
```
if we want to be fully generic, we should not be calling `log.Errorln`. the ErrorHandler function should do that.

once we merge this, you should update internal-apis and add something like 

```go
import lbry.go/api
api.ErrorHandler = func(e error){
log.Errorln(e)
SendToSlack(e)
}
```

tiger5226 replied:
Yeah that is a good point. We should not have a dependency on logrus.

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		errMsg := "Error encoding JSON response: " + err.Error()
+		log.Errorln(errMsg)
+		ErrorHandler(errors.Base(errMsg))
```
when you do this, you're losing the trace of the original error. you should probably change this whole section to

```go
if err != nil {
  ErrorHandler(errors.Prefix("error encoding JSON response", err))
}
```

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,299 @@
+var HeaderSettings *map[string]string
+
+// ErrorHandler Allows specific error logging for the server at specific points.
+var ErrorHandler = func(err error) {}
```
I would do this to ensure that if someone passes their own error handler function, its correct:

```go
type ErrorHandlerFunc func(error)
var ErrorHandler ErrorHandlerFunc = func(error) {}
```

I'd also rename this to LogErrorFunc and LogError, since that's more accurate. The error doesnt need to be handled (that already happens). It just needs to be logged.

tiger5226 replied:
golang knows whether or not the function signature matches. If you don't pass the right signature it will not compile. I didn't see another reason to create a type for it. See screen shot below
<img width="882" alt="screen shot 2018-05-12 at 9 09 44 pm" src="https://user-images.githubusercontent.com/3402064/39962836-e49f1d46-5628-11e8-8341-2f3071facf12.png">

I will renamed the var though, makes sense.

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,299 @@
+		if errors.HasTrace(rsp.Error) && len(trace) == 0 {
+			errorMessage += "\n" + errors.Trace(rsp.Error)
+		}
+		log.Errorln(errorMessage)
```
same as above

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,293 @@
+		LogError(errors.Prefix("Error encoding JSON response: ",err))
+	}
+
+	if rsp.Status >= http.StatusInternalServerError {
```
this if statement should be dropped in favor of logging the error above in LogError

tiger5226 replied:
These are two different errors for different reasons. There could be a case where there is not an encoding error but an error in calling the handler. Leaving for now, as I think it is an oversight.

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,293 @@
+	// 	consoleText += " " + string(queryJSON)
+	// }
+	if success {
+		LogInfo(color.GreenString(consoleText))
```
dont use `color` here. that's up to the LogInfo / LogError implementation.

In fact, LogInfo should probably just be passed the request and response, so it can do whatever it wants. And LogError should be passed request, response, and error. There's no need for us to build this string here. What if another implementation wants different data.

tiger5226 replied:
good point.

lyoshenka commented on api/server.go:
```diff
@@ -0,0 +1,293 @@
+		http.Redirect(w, r, rsp.RedirectURL, rsp.Status)
+		return
+	} else if rsp.RedirectURL != "" {
+		// warn us that we messed up somehow
```
should LogError here. something like "status code indicates a redirect, but RedirectURL is empty"

tiger5226 replied:
I think you misread it. It is checking a range and if inside the range it redirects. otherwise it is checking if the redirect is NOT empty. So I change to be `status code %s does not indicate a redirect, but RedirectURL is non-empty '%s'`

tiger5226 commented:
Can we merge this in now? ( DO NOT DELETE BRANCH ).

lyoshenka commented:
Looks good to me, though to be sure, I'd like to see what internal-apis  looks like using this code. But I think its fine to merge this. If there are any tweaks to make later, I bet they'll be minor.

tiger5226 commented:
All clear I will create an internal-apis branch. This will also fix the travis issue with the webhook where it runs twice. I can make that change here.
